sycl/include/CL/sycl/usm/usm_allocator.hpp

```diff
@@ -30,108 +30,43 @@ template <typename T, usm::alloc AllocKind, size_t Alignment = 0>
 class usm_allocator {
 public:
   using value_type = T;
-  using pointer = T *;
-  using const_pointer = const T *;
-  using reference = T &;
-  using const_reference = const T &;
+  using propagate_on_container_copy_assignment = std::true_type;
+  using propagate_on_container_move_assignment = std::true_type;
+  using propagate_on_container_swap = std::true_type;
 
 public:
   template <typename U> struct rebind {
     typedef usm_allocator<U, AllocKind, Alignment> other;
   };
 
-  usm_allocator() = delete;
-  usm_allocator(const context &Ctxt, const device &Dev)
-      : MContext(Ctxt), MDevice(Dev) {}
-  usm_allocator(const queue &Q)
-      : MContext(Q.get_context()), MDevice(Q.get_device()) {}
-  usm_allocator(const usm_allocator &Other)
-      : MContext(Other.MContext), MDevice(Other.MDevice) {}
-
-  /// Constructs an object on memory pointed by Ptr.
-  ///
-  /// Note: AllocKind == alloc::device is not allowed.
-  ///
-  /// \param Ptr is a pointer to memory that will be used to construct the
-  /// object.
-  /// \param Val is a value to initialize the newly constructed object.
-  template <
-      usm::alloc AllocT = AllocKind,
-      typename std::enable_if<AllocT != usm::alloc::device, int>::type = 0>
-  void construct(pointer Ptr, const_reference Val) {
-    new (Ptr) value_type(Val);
+  usm_allocator() noexcept = delete;
+  usm_allocator(const context &Ctxt, const device &Dev) noexcept
+      : MContext(Ctxt), MDevice(Dev) {
+    static_assert(AllocKind != usm::alloc::device,
```

Comment:
Given this is a type-requirement, it is a bit more clear (both the code and the error message) if that `static_assert` is in the class body. Also avoids the duplication.

Reply:
Where in the class body?